### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by Devoxin**
### Checks and guidelines:

* [x] Have you checked that there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Have you checked that only **single quotes** are used in the code, apart from the doc-strings?
* [x] Have you run a lint program on your code prior to submission?
* [x] Have you checked if `python run_tests.py` returns no errors?

Tests output: ~~i don't need any~~ All tested on rewrite branch anyway.

### Type of change

* [x] Bug fix
* [x] New feature
* [x] Improvement
* [x] Breaking change
* [x] This change is a documentation update

### Describe the changes:

big change, multiple nodes, load balancing, regions, relatively familiar syntax to 2.1.10
+ some bug fixes

### CODE REVIEW

File: lavalink/websocket.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+        self._loop = self._lavalink._loop
+        asyncio.ensure_future(self.connect())
+
+    @property
+    def connected(self):
+        """ Returns whether the websocket is connected to Lavalink. """
+        return self._ws is not None and not self._ws.closed
+
+    async def connect(self):
```
Comment (AndyTempel) [inline lavalink/websocket.py]: Just a comment to me: I'll take a look at threading to run ws on a separate loop.

File: setup.py [hunk 2]
```diff
@@ -14,12 +14,12 @@
...
-    description='A lavalink interface built for discord.py',
+    description='A lavalink interface built for discord.py',  # TODO: Change this if I successfully generic-ify the client
     author='Devoxin',
     author_email='luke@serux.pro',
     url='https://github.com/Devoxin/Lavalink.py',
     download_url='https://github.com/Devoxin/Lavalink.py/archive/{}.tar.gz'.format(version),
     keywords=['lavalink'],
     include_package_data=True,
-    install_requires=['websockets>=5.0.1,<6.1.0', 'aiohttp']
+    install_requires=['aiohttp']
```
Comment (AndyTempel) [inline setup.py]: should add aiodns and cchardet, maybe as an optional lavalink[performance] package or just by default OR just put it somewhere in the readme. https://rickbot.is-not-on.ovh/i/7uLicEhTkmpgZg.png

File: lavalink/websocket.py (hunk 1 above)
Comment (Devoxin) [reply lavalink/websocket.py]: Is this not already achieved by using aiohttp in the first place, where it uses a pool of worker threads? Additionally message receiving and event handling is done within a future which I believe assists asynchronous usage?

File: lavalink/nodemanager.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        self.regions = {
+            'asia': ('hongkong', 'singapore', 'sydney', 'japan', 'southafrica'),
+            'eu': ('eu', 'amsterdam', 'frankfurt', 'russia', 'vip-amsterdam', 'london'),
+            'us': ('us', 'brazil', 'vip-us')
+        }
+
+    def __iter__(self):
+        for n in self.nodes:
+            yield n
+
```
[the remark this answers is not in the document]
Comment (Devoxin) [reply lavalink/nodemanager.py]: `len(node_manager)` would look a bit odd, but I can add if necessary. Makes sense to add support for things like `for node in node_manager`

File: lavalink/playermanager.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+class PlayerManager:
+    def __init__(self, lavalink, player):
+        if not issubclass(player, BasePlayer):
+            raise ValueError('Player must implement BasePlayer or DefaultPlayer.')
+        self._lavalink = lavalink
+        self.players = {}
+        self.default_player = player
+
+    def __iter__(self):
```
Comment (Zenrac) [inline lavalink/playermanager.py]: How about `__len__` here ?
`len(lavalink.players)` looks fine.

File: lavalink/websocket.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        elif op == 'event':
+            await self._handle_event(data)
+        else:
+            log.warning('Received unknown op: {}'.format(op))
+
+    async def _handle_event(self, data: dict):
+        player = self._lavalink.players.get(int(data['guildId']))
+
+        if not player:
+            log.warning('Received event for non-existent player! Node: `{}`, GuildId: {}'.format(self._node.name, data['guildId']))
```
Comment (Zenrac) [inline lavalink/websocket.py]: The problem here is that u know that there's no player but sometime you use it after `player.current` etc 3 lines above.
Moreover it happens when removing a player ?

**Pull request merged**